app/naive_bayes.py

```diff
@@ -0,0 +1,286 @@
+
+        :param op_data: local past operational data object
+        """
+        matrix_link_data = []
+        feature_names = [
+            "Status",
+            "Ability_ID", 
+            "Link_Facts", 
+            "Planner",
+            "Obfuscator",
+            "Adversary_ID",
+            "Adversary_Name",
+            "Command",
+            "Number_Facts",
+            "Visibility_Score",
+            "Executor_Platform",
+            "Executor_Name",
+            "Agent_Protocol",
+            "Trusted_Status",
+            "Agent_Privilege",
+            "Host_Architecture"
+        ]
+        # first row is column names
+        matrix_link_data.append(feature_names)
+
+        for cur_operation in op_data:         
+            dict_agent_paw_to_info = {}
+            for agent in cur_operation.agents:
+                dict_agent_paw_to_info[agent.paw] = [agent.contact, agent.trusted, agent.privilege, agent.architecture]           
+            
+            for cur_link in cur_operation.chain:
+                dict_link_features = {}
+                dict_link_features["Planner"] = cur_operation.planner.name
+                dict_link_features["Obfuscator"] = cur_operation.obfuscator
+                dict_link_features["Adversary_ID"] = cur_operation.adversary.adversary_id
+                dict_link_features["Adversary_Name"] = cur_operation.adversary.name
+                dict_link_features["Ability_ID"] = cur_link.ability.ability_id
+                dict_link_features["Status"] = cur_link.status
+                dict_link_features["Command"] = str(b64decode(cur_link.command).decode('utf-8', errors='ignore').replace('\n', ''))
+                dict_link_features["Number_Facts"] = len(cur_link.used)
+                dict_link_features["Visibility_Score"] = cur_link.visibility.score
+                dict_link_features["Executor_Platform"] = cur_link.executor.platform
+                dict_link_features["Executor_Name"] = cur_link.executor.name
+                
+                agent_paw = cur_link.paw
+                if agent_paw in dict_agent_paw_to_info.keys():
+                    contact_type, trusted_status, privilege, architecture = dict_agent_paw_to_info[agent_paw]
+                    dict_link_features["Agent_Protocol"] = contact_type
+                    dict_link_features["Trusted_Status"] = trusted_status
+                    dict_link_features["Agent_Privilege"] = privilege
+                    dict_link_features["Host_Architecture"] = architecture
+                else:
+                    dict_link_features["Agent_Protocol"] = None
+                    dict_link_features["Trusted_Status"] = None
+                    dict_link_features["Agent_Privilege"] = None
+                    dict_link_features["Host_Architecture"] = None
+
+                dict_link_features["Link_Facts"] = self._get_useful_facts(cur_link)
+
+                link_row = [
+                    dict_link_features["Status"],
```

Comment:
In the same vein as the previous comment, these strings being hard coded may be a bit fragile. The ones earlier in this method are harder to deal with, but at least here, you could probably shorten this block by leveraging the `feature_names` list and list comprehension. You could do something like `link_row = [dict_link_features[feature] for feature in feature_names]` which would building the row a bit more resistant to changes in the features down the line.